src/crimson/net/Dispatcher.h

```diff
@@ -54,6 +55,11 @@ class Dispatcher {
   }
   virtual seastar::future<std::unique_ptr<AuthAuthorizer>>
   ms_get_authorizer(peer_type_t);
+
+  // get the local dispatcher shard if it is accessed by another core
+  virtual Dispatcher* get_local_shard() {
```

Comment:
i have a feeling that we are going to have a melange of the concept of `shared<>` and `Service` here. as you are implying that `Service` should be aware that it's being sharded. the caller site of this method looks like
```c++
seastar::future<> SocketMessenger::start(Dispatcher *disp) {
   return container().invoke_on_all([disp](auto& msgr) {
       return msgr.do_start(disp->get_local_shard());
     });
}
```
but i think the caller of `SocketMessenger::start()` should be responsible for passing the *local* service to to *local* instance of `SocketMessenger`.

Reply:
The major purpose of `get_local_shard()` is to allow any caller holding an abstract `Dispatcher` to get the local shard if the concrete `Dispatcher` class is sharded. While, if the the concrete `Dispatcher` class is not sharded, it can simply reuse the default implementation of `Dispatcher::get_local_shard()` to get the compatible behavior.

The specific example here, is to let `SocketMessenger` to always get a valid pointer regardless of whether the concrete Dispatcher is sharded or not.

Reply:
well, thought there was some alternatives to gap the difference. but let's keep it this way. will come up with another (hopefully better) solution later on.